src/output/mod.rs

```diff
@@ -0,0 +1,479 @@
+use serde::Serialize;
+use std::io::{BufWriter, Write};
+use std::sync::Arc;
+
+#[macro_export]
+macro_rules! output {
+    () => {
+        let now = clocksource::DateTime::now();
+        println!("{}", now.to_rfc3339_opts(clocksource::SecondsFormat::Millis, false));
+    };
+    ($($arg:tt)*) => {{
+        let now = clocksource::DateTime::now();
+        println!("{} {}", now.to_rfc3339_opts(clocksource::SecondsFormat::Millis, false), format_args!($($arg)*));
+    }};
+}
+
+pub fn log(config: &Config, traffic_ratelimit: Option<Arc<Ratelimiter>>) {
+    let mut interval = config.general().interval().as_millis();
+    let mut duration = config.general().duration().as_millis();
+
+    let mut window_id = 0;
+
+    let mut snapshot = Snapshot {
+        prev: HashMap::new(),
+    };
+
+    let mut prev = Instant::now();
+
+    let mut windows_under_target_rate = 0;
+    let mut windows_over_p999_slo = 0;
+
+    let client = !config.workload().keyspaces().is_empty();
+    let pubsub = !config.workload().topics().is_empty();
+
+    while duration > 0 {
+        std::thread::sleep(Duration::from_millis(1));
+
+        interval = interval.saturating_sub(1);
+        duration = duration.saturating_sub(1);
+
+        if interval == 0 {
+            let now = Instant::now();
+            let elapsed = now.duration_since(prev).as_secs_f64();
+            prev = now;
+
+            output!("-----");
+            output!("Window: {}", window_id);
+
+            // output the client stats and retrieve the number of client
+            // requests sent since last snapshot
+            let client_ok = if client {
+                client_stats(&mut snapshot, elapsed)
+            } else {
+                0
+            };
+
+            // output the pubsusb stats and retrieve the number of messages
+            // published since last snapshot
+            let publish_ok = if pubsub {
+                pubsub_stats(&mut snapshot, elapsed)
+            } else {
+                0
+            };
+
+            // the total number of requests + publishes
+            let total_ok = client_ok + publish_ok;
+
+            // a check to determine if we're approximately hitting our target
+            // ratelimit. If not, this will terminate the run.
+            if config.workload().strict_ratelimit() {
+                if let Some(rate) = traffic_ratelimit.as_ref().map(|v| v.rate()) {
+                    if total_ok as f64 / elapsed < 0.95 * rate as f64 {
+                        windows_under_target_rate += 1;
+                    } else {
+                        windows_under_target_rate = 0;
+                    }
+
+                    if windows_under_target_rate > 5 {
+                        break;
+                    }
+                }
+            }
+
+            // a check to determine if we're achieving our p999 SLO (if set). If
+            // not, this will terminate the run.
+            if config.workload().p999_slo() > 0 {
```

Comment:
I see- we probably need an SLO trait? E.g. I can see SLO being declared against an entire vector of percentiles and targets and all have to match, or it will be declared against something other than p99. Better to have a generic `breached()` method.

Reply:
I think this feature needs more thought. I think the rate not hitting ratelimit cutoff is more important for now.

Proposing to remove the SLO check for now and come back with a more fully-considered solution in a future PR.

Reply:
#13 opened for this.